File: backend/app/api/project.py
from fastapi import APIRouter, HTTPException
from typing import List

from app.models.types import ScanConfig, TiltSeries, BatchSaveRequest
from app.state.project_state import project_state

router = APIRouter()


@router.post("/scan")
async def scan_project(config: ScanConfig):
	"""Scan project directory for mdoc files"""
	# Import here to avoid circular dependency
	from app.api.mdoc import parse_mdoc_file
	from app.matcher.cut_match import ImageMatcher
	from pathlib import Path

	try:
		project_state.set_config(config)

		# Build image matcher cache
		matcher = ImageMatcher(
			config.image_dir,
			config.image_prefix_cut,
			config.image_suffix_cut,
		)
		matcher.build_cache()

		# Scan for mdoc files
		mdoc_dir = Path(config.mdoc_dir)
		if not mdoc_dir.exists():
			raise HTTPException(status_code=404, detail=f"mdoc directory not found: {config.mdoc_dir}")

		tilt_series: List[TiltSeries] = []

		for mdoc_file in mdoc_dir.rglob("*.mdoc"):
			try:
				ts = parse_mdoc_file(str(mdoc_file), matcher)
				project_state.add_tilt_series(ts)
				tilt_series.append(ts)
			except Exception as e:
				print(f"Warning: Failed to parse {mdoc_file}: {e}")
				continue

		return {"tiltSeries": tilt_series, "total": len(tilt_series)}

	except HTTPException:
		raise
	except Exception as e:
		raise HTTPException(status_code=500, detail=f"Scan failed: {str(e)}")


@router.get("/status")
async def get_project_status():
	"""Get current project status"""
	return {
		"totalSeries": len(project_state.list_tilt_series()),
		"hasConfig": project_state.config is not None,
		"unsavedCount": len(project_state.frame_overrides),
	}


@router.post("/save_all")
async def save_all():
	"""Save all modified tilt series"""
	try:
		saved_count = 0
		failed_count = 0
		errors = []

		for ts in project_state.list_tilt_series():
			overrides = project_state.get_all_overrides(ts.mdocPath)
			if not overrides:
				continue

			try:
				# Import here to avoid circular dependency
				from app.api.mdoc import batch_save as save_ts
				request = BatchSaveRequest(mdocPath=ts.mdocPath, selections=overrides)
				result = await save_ts(request)
				if result.success:
					saved_count += 1
				else:
					failed_count += 1
					errors.append(f"{ts.id}: {result.message}")
			except Exception as e:
				failed_count += 1
				errors.append(f"{ts.id}: {str(e)}")

		return {
			"success": failed_count == 0,
			"savedCount": saved_count,
			"failedCount": failed_count,
			"errors": errors,
		}

	except Exception as e:
		raise HTTPException(status_code=500, detail=f"Save all failed: {str(e)}")

File: backend/src/routes/project.rs
use std::path::Path;

use axum::Json;
use serde_json::{Value, json};

use crate::matcher::cut_match::ImageMatcher;
use crate::mdoc::parser::parse_mdoc_file;
use crate::models::types::{ScanConfig, TiltSeries};
use crate::state::project_state::PROJECT_STATE;

pub fn router() -> axum::Router {
    axum::Router::new()
        .route("/scan", axum::routing::post(scan_project))
        .route("/status", axum::routing::get(get_project_status))
        .route("/save_all", axum::routing::post(save_all))
}

async fn scan_project(Json(config): Json<ScanConfig>) -> Result<Json<Value>, (axum::http::StatusCode, String)> {
    tracing::info!("[project] Scanning project: {:?}", config);

    PROJECT_STATE.set_config(config.clone()).await;

    let mut matcher = ImageMatcher::new(
        &config.image_dir,
        config.image_prefix_cut as usize,
        config.image_suffix_cut as usize,
    );
    matcher.build_cache();

    let mdoc_dir = Path::new(&config.mdoc_dir);
    if !mdoc_dir.exists() {
        return Err((
            axum::http::StatusCode::NOT_FOUND,
            format!("mdoc directory not found: {}", config.mdoc_dir),
        ));
    }

    let mut tilt_series: Vec<TiltSeries> = Vec::new();

    let walker = walkdir::WalkDir::new(mdoc_dir).max_depth(10);
    for entry in walker.into_iter().filter_map(|e| e.ok()) {
        if entry.file_type().is_file() {
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) == Some("mdoc") {
                match parse_mdoc_file(&path.to_string_lossy(), &matcher) {
                    Ok(ts) => {
                        PROJECT_STATE.add_tilt_series(ts.clone()).await;
                        tilt_series.push(ts);
                    }
                    Err(e) => {
                        tracing::warn!("Failed to parse {}: {}", path.display(), e);
                    }
                }
            }
        }
    }

    Ok(Json(json!({
        "tiltSeries": tilt_series,
        "total": tilt_series.len()
    })))
}

async fn get_project_status() -> Json<Value> {
    let series = PROJECT_STATE.list_tilt_series().await;
    Json(json!({
        "totalSeries": series.len(),
        "hasConfig": true,
        "unsavedCount": 0
    }))
}

async fn save_all() -> Result<Json<Value>, (axum::http::StatusCode, String)> {
    let series = PROJECT_STATE.list_tilt_series().await;
    let mut saved_count = 0;

    for _ts in &series {
        // Selections are managed by frontend; this endpoint just confirms state
        saved_count += 1;
    }

    Ok(Json(json!({
        "success": true,
        "savedCount": saved_count,
        "failedCount": 0,
        "errors": []
    })))
}
